fixes #12436 - as a user I want to deploy vSphere vms on a storage pod

timogoebel opened the pull request:
This PR implements creating and cloning a vm on a datastore cluster instead of a datastore and adds datastore clusters to the api.

Creating on a datastore:
![image](https://cloud.githubusercontent.com/assets/4107560/11186272/34a4d660-8c81-11e5-88ba-7ec4b2396146.png)

Creating on a datastore cluster:
![image](https://cloud.githubusercontent.com/assets/4107560/11186298/482c7288-8c81-11e5-80db-ef7d42630f1f.png)

@brandonweeks: Mind to take a look? What do you think?

timogoebel committed:
fixes #12436 - as a user I want to deploy vSphere vms on a storage pool

timogoebel commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -366,7 +378,8 @@ def clone_vm(raw_args)
         "datastore" => args[:volumes].first[:datastore],
-        "resource_pool" => [args[:cluster], args[:resource_pool]],
+        "storage_pod" => args[:volumes].first[:storage_pod],
+        #"resource_pool" => [args[:cluster], args[:resource_pool]],
```
This broke cloning for me. Have to investigate this.

timogoebel replied:
I could not reproduce this error anymore. I was probably just forgetting something...

timogoebel commented on app/assets/javascripts/compute_resource.js:
```diff
@@ -252,3 +252,26 @@ function vsphereGetResourcePools(item) {
+function vsphereStoragePodSelected(item) {
+  selected = $(item).val();
+  dropdown = $(item).closest('select');
+  datastore = $(item).parentsUntil('.fields').parent().find('#host_compute_attributes_volumes_attributes_0_datastore');
```
This only works for the first volume. Any hints how to fix this?

dLobatog replied:
How about matching a regex instead of finding an ID directly? I understand the `0` is the problem right?

Also would this work on a separate file? It's part of compute_resource but I think it's a bit easier to understand conceptually if we stick all the VMware logic in its own file, 😶

timogoebel commented:
This requires changes to fog, see https://github.com/fog/fog-vsphere/pull/6

timogoebel commented on app/views/compute_resources_vms/form/vmware/_volume.html.erb:
```diff
@@ -1,3 +1,4 @@
+<%= selectable_f f, :storage_pod, vsphere_storage_pods(compute_resource), { :include_blank => _('Deploy VM on selected datastore') }, :class => "span5 without_select2", :label => _("Datastore Cluster"), :label_size => "col-md-3", :onchange => 'vsphereStoragePodSelected(this)' if new_host and vsphere_storage_pods(compute_resource).any? %>
 <%= selectable_f f, :datastore, vsphere_datastores(compute_resource), { }, :class => "span5 without_select2", :label => _("Data store"), :label_size => "col-md-3" %>
```
This should be disabled when a storage pod is selected. Any ideas how this can be done as there might be multiple volumes...?

timogoebel replied:
Fixed via jquery, see vsphereStoragePodLoad()

dLobatog commented on app/controllers/api/v2/compute_resources_controller.rb:
```diff
@@ -15,7 +15,7 @@ class ComputeResourcesController < V2::BaseController
                                               :available_clusters, :available_flavors, :available_folders,
                                               :available_networks, :available_resource_pools, :available_security_groups, :available_storage_domains,
-                                              :available_zones]
+                                              :available_zones, :available_storage_pods]
```
This is becoming a bit unwieldy, if you have any ideas to refactor this out that'd be welcome

timogoebel replied:
I do not have any ideas how that could be refactored. Should I raise a Redmine for this?

dLobatog commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -68,6 +68,18 @@ def datastores(opts = {})
 
+    def storage_pods(opts = {})
+      if opts[:storage_pod]
+        dc.storage_pods.get(opts[:storage_pod]) rescue {}
```
Mind to rescue the appropriate possible Fog / network exceptions? This can hide errors about the connection between Foreman and vSphere that can be hard to debug.

dLobatog commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -68,6 +68,18 @@ def datastores(opts = {})
+      if opts[:storage_pod]
+        dc.storage_pods.get(opts[:storage_pod]) rescue {}
+      else
+        name_sort(dc.storage_pods.all()) rescue []
```
Same thing

dLobatog commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -68,6 +68,18 @@ def datastores(opts = {})
+      end
+    end
+
+    def available_storage_pods(storage_pod = nil)
```
What's an `unavailable` storage pod? A full one? I'm just wondering if we need the `available` prefix here, or if you want to make checks to make sure they are available indeed :smile:

timogoebel replied:
Actually I just followed the pattern established. All methods like storage_domains always start with 'available_'...

dLobatog commented on app/helpers/compute_resources_vms_helper.rb:
```diff
@@ -84,6 +84,15 @@ def datastore_stats(datastore)
+  end
+
+  def storage_pod_stats(pod)
+    return pod.name unless pod.freespace && pod.capacity
```
Why would .freespace and .capacity not be available on a pod?

dLobatog commented:
Nice work @timogoebel , it's lacking tests and some testing, but it's a good step on the right direction towards better VMware integration. Thanks!

timogoebel commented:
I updated the code a bit. Let's wait for tests...

timogoebel commented:
@dLobatog : What kind of tests do you want me to add?

timogoebel commented:
Katello tests are failing. No idea, why.

shlomizadok commented:
They sometimes randomly fail. [test]
On Nov 26, 2015 18:15, "Timo Goebel" notifications@github.com wrote:

> Katello tests are failing. No idea, why.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/theforeman/foreman/pull/2915#issuecomment-159951518.

dLobatog commented:
@timogoebel At least one test for the API/v2 controller would help. You can use Fog.mock to get sample responses from Fog.

timogoebel commented:
@dLobatog : Thanks for reviewing, added the tests. Hope you had something like this in mind.

dLobatog commented on test/functional/api/v2/compute_resources_controller_test.rb:
```diff
@@ -230,6 +235,19 @@ def teardown
+
+    get :available_storage_pods, { :id => compute_resources(:vmware).to_param, :storage_pod => 'test_vmware_cluster' }
+    assert_response :success
+    available_storage_domains = ActiveSupport::JSON.decode(@response.body)
```
This variable should be `available_storage_pods`, Rubocop it's complaining it's not being used & the test is failing

timogoebel replied:
Oh darn, how could I have missed that. Thanks, fixed! :-)

dLobatog commented:
@timogoebel Set to WoC as tests are failing - I'll rereview after that :smile: thank you again!

timogoebel commented:
@dLobatog : Tests are probably failing on jenkins, because we need the fog patches for this. Let's wait for a release, then we can continue to work on this.

timogoebel commented:
Rebased on current develop.

dLobatog commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -1,5 +1,6 @@
 require 'fog_extensions/vsphere/mini_servers'
 require 'foreman/exception'
+require 'rbvmomi'
```
I don't think this require is needed, as long as fog-vsphere has been required this should work. Notice how we use RbVmomi in this class w/o requiring it, line 262 for instance.

dLobatog commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -69,6 +70,26 @@ def datastores(opts = {})
+      else
+        begin
+          name_sort(dc.storage_pods.all())
+        rescue RbVmomi::VIM::InvalidArgument
```
Kinda confused at these rescues, why does one of them return a hash & the other returns an array? I'd assume the first one would return an array containing the storage pod when you pass the option, and the second one would return an array of all storage pods.

dLobatog commented:
Made 2 small comments I'd need to understand better, I also need to test provisioning with this, but overall :+1:, thanks! One step closer to merging :smile:

timogoebel commented:
@dLobatog : Thanks for reviewing!
I removed the require and added code comments to the rescues. I hope, this makes it more clear.
Basically, if your vCenter does not support storage pods, we need to return an empty list (e.g. empty array). So you got it right regardless of the comments.

However, I found a bug in fog-vsphere, which breaks storage pod functionality. I have created a pr to fix this upstream. https://github.com/fog/fog-vsphere/pull/22

timogoebel commented:
@dLobatog: fog-vsphere 0.5.0 just got released and fixes the issue. I have updated the bundler dependency it in this pr.

angystardust commented:
Any hope to see this one in 1.11 RC, @dLobatog ? :pray:

timogoebel commented:
@dLobatog : Sorry for being a pain, how is the status of this? We use the code in production for some time now and it works great. So I figured: Why keep the rest of the foreman community from using it as well? :-P

dLobatog commented:
@timogoebel Sorry for the wait, it took me a while to get ESXi running on KVM and also vSphere to set up the storage pod. Seemed to work well so.. :bomb:

dLobatog commented:
Merged as d46cc6931b7ab1ec0860a4c389c07ae5180cf768, thanks @timogoebel!
